src/docker-extract.c

```diff
@@ -0,0 +1,342 @@
+                            wh_path_rootfs);
+        retval = unlink(wh_path_rootfs);
+    }
+
+    free(prefix_str);
+    free(suffix_str);
+    free(wh_path);
+    free(wh_path_rootfs);
+
+    return retval;
+}
+
+/* apply_whiteouts
+ *  Process tarfile and apply any aufs opaque/whiteouts on rootfs_dir
+ */
+int apply_whiteouts(char *tarfile, char *rootfs_dir) {
+    int retval = 0;
+    int errcode = 0;
+
+    struct archive *a;
+    struct archive_entry *entry;
+
+    a = archive_read_new();
+#if ARCHIVE_VERSION_NUMBER <= 3000000
+    archive_read_support_compression_all(a);
+#else
+    archive_read_support_filter_all(a);
+#endif
+    archive_read_support_format_all(a);
+    retval = archive_read_open_filename(a, tarfile, 10240);
+    if (retval != ARCHIVE_OK) {
+        return (1);
+    }
+
+    while (archive_read_next_header(a, &entry) == ARCHIVE_OK) {
+
+        const char *pathname = archive_entry_pathname(entry);
+
+        if (strstr(pathname, "/.wh..wh..opq")) {
+            singularity_message(DEBUG, "Opaque Marker %s\n", pathname);
+            errcode = apply_opaque(pathname, rootfs_dir);
+            if (errcode != 0) {
+                break;
+            }
+        } else if (strstr(pathname, "/.wh.")) {
+            singularity_message(DEBUG, "Whiteout Marker %s\n", pathname);
+            errcode = apply_whiteout(pathname, rootfs_dir);
+            if (errcode != 0) {
+                break;
+            }
+        }
+    }
+#if ARCHIVE_VERSION_NUMBER <= 3000000
+    retval = archive_read_finish(a);
+#else
+    retval = archive_read_free(a);
+#endif
+    if (retval != ARCHIVE_OK){
+        singularity_message(ERROR, "Error freeing archive\n");
+        ABORT(255);
+    }
+
+    return errcode;
+}
+
+/* See  https://github.com/libarchive/libarchive/wiki/Examples#A_Complete_Extractor */
+static int copy_data(struct archive *ar, struct archive *aw) {
+    int r;
+    const void *buff;
+    size_t size;
+    int64_t offset;
+
+    for (;;) {
+        r = archive_read_data_block(ar, &buff, &size, &offset);
+        if (r == ARCHIVE_EOF) {
+            return (ARCHIVE_OK);
+        }
+        if (r < ARCHIVE_OK) {
+            return (r);
+        }
+        r = archive_write_data_block(aw, buff, size, offset);
+        if (r < ARCHIVE_OK) {
+            singularity_message(ERROR, "tar extraction error: %s\n", archive_error_string(aw));
+            return (r);
+        }
+    }
+}
+
+/* extract_tar
+ *  Extract a tar file to rootfs_dir using libarchive. Handles compression.
+ *  Exclude any .wh. whiteout files, and device/pipe/fifo entries
+ *
+ * See https://github.com/libarchive/libarchive/wiki/Examples#A_Complete_Extractor
+ */
+int extract_tar(const char *tarfile, const char *rootfs_dir) {
+    int retval = 0;
+    struct archive *a;
+    struct archive *ext;
+    struct archive_entry *entry;
+    int flags;
+    int r;
+    char *orig_dir;
+    const char *pathname;
```

Comment:
Actually I think that pathname should be declared at the beginning of the for loop where it is used since that is the only scope it is needed for.